### PR TITLE
CORE-16201 Multi-Source Event Mediator - Message Processor Task

### PR DESCRIPTION
**Pull request by mbrkic-r3**
Message processor task that enables Multi-Source Event Mediator to process events in parallel using `TaskManger`. It takes events with the same key and associated state, invokes message processor and returns output events and updated state.

### CODE REVIEW

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    private val stateManager: StateManager,
+    private val serializer: CordaAvroSerializer<Any>,
+    private val stateDeserializer: CordaAvroDeserializer<S>,
+) {
+
+    /**
+     * Creates an updated [State] or a new one if there was no previous version.
+     *
+     * @param key Event's key.
+     * @param persistedState State bieing updated.
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt]: bieing --> being

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            persistedState?.metadata ?: Metadata()
+        )
+    }
+
+    /**
+     * Persists states of [ProcessorTask] results.
+     *
+     * @param processorTaskResults [ProcessorTask] results with updated states.
+     * @return The latest states in case persistence failed due to conflict (state being updated by another process in
+     * the meanwhile).
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt]: nit -> in the meantime

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import net.corda.libs.statemanager.api.Metadata
+import net.corda.libs.statemanager.api.State
+import net.corda.libs.statemanager.api.StateManager
+
+/**
+ * Helper for working with [StateManager], used by [MultiSourceEventMediatorImpl].
+ */
+class StateManagerHelper<K : Any, S : Any, E : Any>(
+    private val stateManager: StateManager,
+    private val serializer: CordaAvroSerializer<Any>,
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt]: should this be called stateSerializer so name is consistent?

**Review by jujoramos — DISMISSED** [review]
LGTM

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+        `when`(stateManagerHelper.createOrUpdateState(any(), anyOrNull(), anyOrNull())).thenReturn(
+            mock()
+        )
+    }
+
+    @Test
+    fun `successfully processes events`() {
+
+        val persistedSate: State? = null
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt]: persistedSate --> persistedState

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            processor,
+            stateManagerHelper,
+        )
+
+        val result = task.call()
+
+        verify(processor, times(eventRecords.size)).onNext(stateCaptor.capture(), eventCaptor.capture())
+        val capturedStates = stateCaptor.allValues
+        val expectedInputStates = listOf(null, StateType(0), StateType(1))
+        assertEquals(expectedInputStates, capturedStates)
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt]: nit -> I'd make the naming line up so it's like:
expectedInputStates & capturedInputStates

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        val capturedEvents = eventCaptor.allValues
+        assertEquals(eventRecords, capturedEvents)
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt]: nit: these are both lists of records so I might rename to something like 
eventRecords & capturedEventRecords 
or 
expectedEventRecords & capturedEventRecords

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/StateManagerHelperTest.kt [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    companion object {
+        private const val TEST_KEY = "key"
+        private val TEST_STATE_VALUE = StateType(1)
+    }
+
+    private data class StateType(val id: Int)
+    private class EventType
+
+    private val stateManager = mock<StateManager>()
+    private val serializer = mock<CordaAvroSerializer<Any>>()
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/StateManagerHelperTest.kt]: stateSerializer

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/StateManagerHelperTest.kt [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            serialized(value)
+        }
+    }
+
+    private fun serialized(value: Any) = value.toString().toByteArray()
+
+    @Test
+    fun `successfully creates new state`() {
+
+        val persistedSate: State? = null
```
Comment (emilybowe) [inline libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/StateManagerHelperTest.kt]: persistedState

**Review by emilybowe — DISMISSED** [review]
LGTM, left a few small comments

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt (hunk 6 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt]: fixed

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/StateManagerHelperTest.kt (hunk 8 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/StateManagerHelperTest.kt]: fixed

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/StateManagerHelperTest.kt (hunk 7 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/StateManagerHelperTest.kt]: fixed

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt (hunk 5 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt]: fixed

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt (hunk 4 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/mediator/ProcessorTaskTest.kt]: fixed

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt (hunk 3 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt]: renamed

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt (hunk 2 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt]: fixed

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt (hunk 1 above)
Comment (mbrkic-r3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/mediator/StateManagerHelper.kt]: fixed

**Review by emilybowe — APPROVED** [review]
LGTM

**Pull request merged**